sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsublite/PerPartitionSdf.java

```diff
@@ -0,0 +1,139 @@
+      Duration maxSleepTime,
+      SerializableBiFunction<Partition, Offset, PullSubscriber<SequencedMessage>> subscriberFactory,
+      SerializableFunction<Partition, Committer> committerFactory,
+      SerializableSupplier<Sleeper> sleeperSupplier,
+      SerializableFunction<Partition, InitialOffsetReader> offsetReaderFactory,
+      SerializableBiFunction<
+              Partition, OffsetRange, RestrictionTracker<OffsetRange, OffsetByteProgress>>
+          trackerFactory) {
+    this.maxSleepTime = maxSleepTime;
+    this.sleepTimeRemaining = maxSleepTime;
+    this.subscriberFactory = subscriberFactory;
+    this.committerFactory = committerFactory;
+    this.sleeperSupplier = sleeperSupplier;
+    this.offsetReaderFactory = offsetReaderFactory;
+    this.trackerFactory = trackerFactory;
+  }
+
+  private List<SequencedMessage> doPoll(PullSubscriber<SequencedMessage> subscriber)
+      throws Exception {
+    Sleeper sleeper = sleeperSupplier.get();
+    while (sleepTimeRemaining.isLongerThan(Duration.ZERO)) {
+      List<SequencedMessage> messages = subscriber.pull();
+      if (!messages.isEmpty()) {
+        return messages;
+      }
+      Duration sleepTime =
+          Collections.min(ImmutableList.of(sleepTimeRemaining, Duration.millis(50)));
+      sleepTimeRemaining = sleepTimeRemaining.minus(sleepTime);
+      sleeper.sleep(sleepTime.getMillis());
+    }
+    return ImmutableList.of();
+  }
+
+  @ProcessElement
+  public ProcessContinuation processElement(
+      RestrictionTracker<OffsetRange, OffsetByteProgress> tracker,
+      @Element Partition partition,
+      OutputReceiver<SequencedMessage> receiver)
+      throws Exception {
+    logger.atInfo().log("Starting processing for partition " + partition);
+    sleepTimeRemaining = maxSleepTime;
+    Committer committer = committerFactory.apply(partition);
+    committer.startAsync().awaitRunning();
+    try (PullSubscriber<SequencedMessage> subscriber =
+        subscriberFactory.apply(partition, Offset.of(tracker.currentRestriction().getFrom()))) {
+      while (true) {
+        List<SequencedMessage> messages = doPoll(subscriber);
+        // We polled for as long as possible, yield to the runtime to allow it to reschedule us on
+        // a new task.
+        if (messages.isEmpty()) {
+          logger.atInfo().log("Yielding due to timeout on partition " + partition);
+          return ProcessContinuation.resume();
+        }
+        Offset lastOffset = Offset.of(Iterables.getLast(messages).getCursor().getOffset());
+        long byteSize = messages.stream().mapToLong(SequencedMessage::getSizeBytes).sum();
+        if (tracker.tryClaim(OffsetByteProgress.of(lastOffset, byteSize))) {
+          messages.forEach(
+              message ->
+                  receiver.outputWithTimestamp(
+                      message, new Instant(Timestamps.toMillis(message.getPublishTime()))));
+          committer.commitOffset(Offset.of(lastOffset.value() + 1)).get();
```

Comment:
What'e the effect of committing offset? Are we able to read from that offset again if it's committed?

Reply:
It updates internal metrics and changes a stored integer value. It has no effect on where the current I/O reads from or what messages can be read.

Reply:
This does actually affect where the I/O will start on clean restarts. Where this should be depends on how the beam model / dataflow handles drains for SDF pipelines. Do you have any information on this?

Reply:
You can either use [bundle finalization](https://beam.apache.org/documentation/programming-guide/#bundle-finalization) to commit offsets, which is best-effort. This is similar to how CheckpointMark of UnboundedSource works. Or you can do something similar to Kafka [CommitOffset](https://github.com/apache/beam/blob/master/sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaCommitOffset.java) transform.

Reply:
Done.